### DISCUSSION TITLE
How does the ListVector work within duckdb for creating +++1D nested Lists for an extension

### DISCUSSION THREAD

**Discussion by HackettJP** [duckdb/duckdb#7042] [Q&A] [answered]
Hi,

I'm trying to enhance the postgres_scanner C version extension, to support more than 1D nested arrays, and just need some Q&A.
From my side my of workings I can spit out the composite , correctly step through and decode the headers and the data-blocks, OID, ndims, dims, upper lower dimensions not much info out there on how to decode.

single column data in postgresql  

{{{NULL,NULL},{TWO,2},{THREE,NULL}},{{FOUR,4},{FIVE,5},{SIX,6}}} 

total field items in that row discounting brackets  : 12
nattrs   flag        : 0  (1) means non-composite 
size                                    : 71
first_tup                               : 1
ndims                                   : 3
nulls flag                             : 1
elemoid                                 : 25
dimension decode start outside in 2 primary arrays with 3 nested arrays within and 2 elements within each array.
 Dimensions : 0  no of elements in that dimension dims value  : 2 , lbound value : 1 
 Dimensions : 1  no of elements in that dimension dims value  : 3 , lbound value : 1 
 Dimensions : 2  no of elements in that dimension dims value  : 2 , lbound value : 1 

output decode.
 [ NULL  NULL   [  TWO    2   [  THREE  NULL  [    FOUR    4  [    FIVE    5  [   SIX    6 ]

rather build my own data-model on how to store Nested dimensions of no limits i paused, and work out how to translate this into duckdb nested listed formats.

within postgres_scanner.c extension 
 LogicalTypeId::LIST:  it can only deal with { } simple array even a {{}} it does not detect correctly (errors)  does a '\0' buffer and {{{}}} seg faults.

just want to know how to use and correctly ingest the data.

ListVector::SetListSize, ListType::GetChildType, FlatVector::Validity, ListVector::Reserve anything else missing.

is it a B-Tree the data model that the duckdb LISTS are used against.
PR some time ago, this is what you use for VARCHAR i think  for starters

struct string_t {
    uint32_t length;  // 4bytes length of string in unsigned 32 bits
    union {
        char data[12];  // stored directly in here if less than 12 bytes
        struct {
            char prefix[4]; // store the first 4 bytes of the string 
            char *data;   /// pointer to entire data if more than 12 bytes.
        };
    };
};

thks

James

**Comment by Tishj** [comment]
For a LIST the Vector uses `list_entry_t`s as data to keep track of every lists length.
```c++
struct list_entry_t {
	uint64_t offset;
	uint64_t length;
};
```

Next to that data, the Vector also has an auxiliary buffer, which is a child Vector.

In the case of multi dimensional LISTs, these are also list Vectors, except for the last level - that has a regular Vector containing all the data for the column.

The offsets and length of the `list_entry_t`s relate to the bottom-level Vector, the one that actually contains the physical data of the column.

**Reply by HackettJP** [reply]
thks
can you any chance describe a bit more on the multi dimensional lists is what i'm trying to implement for Postgresql . 
Dirty Quick steps for a using list Vectors (duckdb own interpretation  of C++ vectors dynamic creation). get confusing when i know of 4-5 means of  vector terminology's (programming, maths, engineering) 
 
I had a look for test-code inside the duckdb src tree and could not see any examples that use the internals. mostly via front end api's 
James

**Reply by Tishj** [reply]
```
CONSTANT INTEGER[][]: 1 = [ [[5, 3], [], [1, 2, 3, 4, 5, 5], [NULL], NULL, [NULL, 1]]]

Vector (LIST):
	data:
		[0] = {
			offset = 0,
			length = 6
		}
	child:
		Vector (LIST):
			data:
				[0] = {
					offset = 0,
					length = 2
				},
				[1] = {
					offset = 2,
					length = 0
				},
				[2] = {
					offset = 2,
					length = 6
				},
				[3] = {
					offset = 8,
					length = 1
				},
				[4] = {
					offset = <uninitialized>,
					length = <uninitialized>
				},
				[5] = {
					offset = 9,
					length = 2
				}
			child:
				Vector (INTEGER):
					data:
						[ 5, 3, 1, 2, 3, 4, 5, 5, <uninitialized>, <uninitialized>, 1]
```
The validity vector (every Vector has one) will let you know if it's null (invalid), these correspond to the \<unitialized\> values

**Reply by HackettJP** [reply, marked as answer]
thks
